gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+| Var-Int
+| NumHeaders
+| uint64
+| Hash to stop at.
+|-
+| NumHeaders * 32
+| HeaderHashes
+| [NumHeaders][32]byte
+| Slice of filter headers.
+|}
+
+=== Compact Filters ===
+
+The last set of messages we introduce are for ''fetching'' the compact filters
+themselves. Light clients can use these two messages to request a compact
+filter for a particular block hash.
+
+The <code>getcfilter</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| 32
+| BlockHash
+| [32]byte
+| Block hash of the Bitcoin block for which the client wishes to fetch a filter.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter requested.
+|}
+
+The <code>cfilter</code> message is defined as follows:
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| 32
+| BlockHash
+| [32]byte
+| Block hash of the Bitcoin block for which the filter is being returned.
+|-
+| 1
+| FilterType
+| byte
+| Byte identifying the type of filter being returned.
+|-
+| Var-Int
+| NumFilterBytes
+| uint64
+| A variable length integer encoding the number of bytes of the filter in the following field.
+|-
+| NumFilterBytes
+| FilterBytes
+| [NumFilterBytes]byte
+| The raw compressed compact filter for this block.
+|}
+
+The <code>BlockHash</code> field is included in both messages as this allows
+easily matching requests against responses, as the responses aren't
+self-identifying like block headers are (via own hash).
+
+The parameters <code>N</code> (the number of elements in the filter) and
+<code>P</code> (<code>1 << false_positive_rate</code>) are required by the light
+client in order to properly incrementally decode, query, and validate
+(reconstruct from Bitcoin block) a compact filter. The parameter <code>N</code>
+cannot be known ahead of time, therefore we define the serialization of a
+compact filter of type <code>0x00</code> and <code>0x01</code> as:
+<pre>
+N || raw_filter_bytes
+</pre>
+where <code>N</code> is serialized as a 32-bit big-endian integer.
+
+However, there exists a special case of a <code>null</code> filter. This this
+case an empty byte slice is transmitted rather than consuming
+<code>4-bytes</code> to encode the size of zero.
+
+However, as the parameter <code>P</code> ''must'' be globally agreed upon (for a
+particular filter type), we define this value ''statically'' for filter types:
+<code>0x00</code> and <code>0x01</code>. For the two aforementioned filter types,
+the false positive rate has been chosen to be: <code>20</code>, meaning the
+parameter <code>P</code> is: <code>2^20</code>, meaning <code>fp=20</code>.
+This value was chosen as during simulations it was the value that minimized the
+bandwidth utilized by the expected number of blocks downloaded due to false
+positives, and the bandwidth used to download the filters themselves. The code along with a demo used for the paramter tuning can be found [here]
+
+
+=== Protocol Version Bump ===
```

Comment:
I do not believe this is neccessary? There is a new service bit, which should be sufficient.

Reply:
AFAIK, all new extensions to the p2p protocol have been accompanied by a protocol version bump? (or at least ones that add _new_ messages)